api/src/application/user/user-service.js

```diff
@@ -30,49 +31,73 @@ module.exports = (bookshelf, emailService, renderTemplate) => {
-        return emailService.send({
-          to: user.get('email'),
-          subject: `${appName} Invitation`,
-          html: emailBody,
-        });
+        }).then(emailBody => {
+          return emailService.send({
+            to: user.get('email'),
+            subject: `${appName} Invitation`,
+            html: emailBody,
+          }).then(data => {
+            return data;
+          }).catch(error => {
+            Boom.badImplementation('Something went wrong', error);
+          });
+        }).catch(error => {
+          Boom.badImplementation('Something went wrong', error);
+        });  
+      }).catch(error => {
+        Boom.badImplementation('Something went wrong', error);
       });
     },
 
-    resendUserInvite(userId, appName, hoursTillExpiration) {
+    resendUserInvite(userId, appName, clientId, redirect_uri, scope, hoursTillExpiration) {
       return bookshelf.model('user').where({ id: userId }).fetch().then(user => {
         if (!user) {
           return Boom.notFound();
         }
+        return clientService.findByRedirectUriAndClientId(payload.client_id, payload.redirect_uri).then(clients => {
+          if (clients.models.length === 0) {
+            return Boom.badData('The provided redirect uri is invalid for the given client id.');
+          }
+          return self.sendInvite(user, appName, clientId, redirect_uri, scope, hoursTillExpiration).then(() => user);
+        });
 
-        return self.sendInvite(user, appName, hoursTillExpiration).then(() => user);
       });
     },
 
     inviteUser(payload) {
       let createdUser;
-      return self.create(
-        payload.email,
-        uuid.v4(),
-        {
-          app_metadata: payload.app_metadata || {},
-          profile : payload.profile || {}
+      return clientService.findByRedirectUriAndClientId(payload.client_id, payload.redirect_uri).then(clients => {
+        if (clients.models.length === 0) {
+          return Boom.badData('The provided redirect uri is invalid for the given client id.');
         }
-      ).then(user => {
-        createdUser = user;
-        return self.sendInvite(
-          user,
-          payload.app_name,
-          payload.hours_till_expiration
-        );
-      }).then(() => createdUser);
+        return self.create(
+          payload.email,
+          uuid.v4(),
+          {
+            app_metadata: payload.app_metadata || {},
+            profile : payload.profile || {}
+          }
+        ).then(user => {
+          createdUser = user;
+          return self.sendInvite(
+            user,
+            payload.app_name,
+            payload.client_id,
+            payload.redirect_uri,
+            payload.scope,
+            payload.hours_till_expiration
+          );
+        }).then(() => createdUser).catch(error => Boom.badImplementation('Something went wrong', error));
+      }).catch(error => {
+        Boom.badImplementation('Something went wrong', error);
+      });
```

Comment:
This `.then()` chain should be un-nested, like so:

``` javascript
return clientService.findByRedirectUriAndClientId()
  .then(clients => {
    // your validation logic
  })
  .then(() => self.create(
    payload.email,
    uuid.vr(),
    {
      app_metadata: payload.app_metadata || {},
      profile : payload.profile || {}
    }
  ))
  // etc
```

Also the two `.catch()` calls shouldn't be necessary